zio-http/src/main/scala/zio/http/model/Header.scala

```diff
@@ -0,0 +1,4453 @@
+import java.nio.charset.{Charset, StandardCharsets, UnsupportedCharsetException}
+import java.time.format.DateTimeFormatter
+import java.time.{ZoneOffset, ZonedDateTime}
+import java.util.Base64
+
+import scala.annotation.tailrec
+import scala.collection.mutable
+import scala.util.control.NonFatal
+import scala.util.matching.Regex
+import scala.util.{Either, Failure, Success, Try}
+
+import zio._
+
+import zio.http._
+import zio.http.model.headers.HeaderNames
+
+sealed trait Header {
+  type Self <: Header
+  def self: Self
+
+  def headerType: Header.HeaderType.Typed[Self]
+  def headerName: String    = headerType.name
+  def renderedValue: String = headerType.render(self)
+
+  private[http] def headerNameAsCharSequence: CharSequence    = headerName
+  private[http] def renderedValueAsCharSequence: CharSequence = renderedValue
+
+  lazy val untyped: Header.Custom = Header.Custom(headerName, renderedValue)
+}
+
+object Header {
+  sealed trait HeaderType {
+    type HeaderValue
+
+    def name: String
+
+    def parse(value: String): Either[String, HeaderValue]
+
+    def render(value: HeaderValue): String
+  }
+
+  object HeaderType {
+    type Typed[HV] = HeaderType { type HeaderValue = HV }
+  }
+
+  final case class Custom(customName: CharSequence, value: CharSequence) extends Header {
+    override type Self = Custom
+    override def self: Self = this
+
+    override def headerType: HeaderType.Typed[Custom] = new Header.HeaderType {
+      override type HeaderValue = Custom
+
+      override lazy val name: String = self.customName.toString
+
+      override def parse(value: String): Either[String, HeaderValue] = Right(Custom(self.customName, value))
+
+      override def render(value: HeaderValue): String = value.value.toString
+    }
+
+    private[http] override def headerNameAsCharSequence: CharSequence    = customName
+    private[http] override def renderedValueAsCharSequence: CharSequence = value
+
+    override def hashCode(): Int = {
+      var h       = 0
+      val kLength = customName.length()
+      var i       = 0
+      while (i < kLength) {
+        h = 17 * h + customName.charAt(i)
+        i = i + 1
+      }
+      i = 0
+      val vLength = value.length()
+      while (i < vLength) {
+        h = 17 * h + value.charAt(i)
+        i = i + 1
+      }
+      h
+    }
+
+    override def equals(that: Any): Boolean = {
+      that match {
+        case Custom(k, v) =>
+          def eqs(l: CharSequence, r: CharSequence): Boolean = {
+            if (l.length() != r.length()) false
+            else {
+              var i     = 0
+              var equal = true
+
+              while (i < l.length()) {
+                if (l.charAt(i) != r.charAt(i)) {
+                  equal = false
+                  i = l.length()
+                }
+                i = i + 1
+              }
+              equal
+            }
+          }
+
+          eqs(self.customName, k) && eqs(self.value, v)
+
+        case _ => false
```

Comment:
Maybe unrealistic, but what if the user extract a header value as `Custom`, then compare it with a modeled one with the same name?

Reply:
Those are different case classes so equality won't be true. But on the `Headers`/`HeaderGetter`/`HeaderModifier` level it is handled so even though all headers from a Netty request or response are `Header.Custom` (not parsed), you can get a typed header value and then it gets parsed, also when modifying headers it does not matter whether it is in typed or unencoded form in the end the last one per key gets encoded.